File: Backtracking/79_WordSearch.py
class Solution(object):

    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        if word == "":
            return False
        len1 = len(board)
        len2 = len(board[0])
        for i in range(len1):
            for j in range(len2):
                if board[i][j] == word[0]:
                    board[i][j] = '*'
                    for p in [-1, 1]:
                        r = self.dosearch(board, word, i + p, j, 1)
                        if r:
                            return True
                        r = self.dosearch(board, word, i, j + p, 1)
                        if r:
                            return True
                    board[i][j] = word[0]
        return False

    def dosearch(self, b, s, i, j, pos):
        if pos == len(s):
            return True
        if i < 0 or j < 0 or i >= len(b) or j >= len(b[0]) or s[pos] != b[i][j]:
            return False
        b[i][j] = '*'
        for p in [-1, 1]:
            r = self.dosearch(b, s, i + p, j, pos+1)
            if r:
                return True
            r = self.dosearch(b, s, i, j + p, pos+1)
            if r:
               return True
        b[i][j] = s[pos]
        return False

File: Backtracking/79_WordSearch.java
public class Solution {

    public boolean exist(char[][] board, String word) {
        int len1 = board.length;
        int len2 = board[0].length;
        char targ = word.charAt(0);
        for (int i=0; i<len1; i++) {
            for (int j=0; j<len2; j++) {
                if (board[i][j] == targ) {
                    board[i][j] = '*';
                    boolean r = search(board, word, i-1, j, 1);
                    if (r) {
                        return true;
                    }
                    r = search(board, word, i+1, j, 1);
                    if (r) {
                        return true;
                    }
                    r = search(board, word, i, j-1, 1);
                    if (r) {
                        return true;
                    }
                    r = search(board, word, i, j+1, 1);
                    if (r) {
                        return true;
                    }
                    board[i][j] = targ;
                }
            }
        }
        return false;
    }

    private boolean search(char[][] b, String word, int i, int j, int pos) {
        if (pos == word.length()) {
            return true;
        }
        if (i<0 || j<0 || i >= b.length || j >= b[0].length || b[i][j] != word.charAt(pos)) {
            return false;
        }
        b[i][j] = '*';
        boolean r = search(b, word, i-1, j, pos+1);
        if (r) {
            return true;
        }
        r = search(b, word, i+1, j, pos+1);
        if (r) {
            return true;
        }
        r = search(b, word, i, j-1, pos+1);
        if (r) {
            return true;
        }
        r = search(b, word, i, j+1, pos+1);
        if (r) {
            return true;
        }
        b[i][j] = word.charAt(pos);
        return false;
    }

}